src/app/OperationalDeviceManager.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/OperationalDeviceManager.h>
+#include <platform/CHIPDeviceLayer.h>
+
+namespace {
+
+struct ConnectionErrorContext
+{
+    chip::OperationalDeviceProxy * device;
+    CHIP_ERROR error;
+};
+
+} // namespace
+
+namespace chip {
+
+CHIP_ERROR OperationalDeviceManager::AcquireDevice(PeerId peerId, Callback::Callback<OnDeviceConnected> * onConnection,
+                                                   Callback::Callback<OnDeviceConnectionFailure> * onFailure)
+{
+    Dnssd::ResolvedNodeData resolutionData;
+
+    bool nodeIDWasResolved = (mConfig.dnsCache != nullptr && mConfig.dnsCache->Lookup(peerId, resolutionData) == CHIP_NO_ERROR);
+
+    OperationalDeviceProxy * device = FindActiveDevice(peerId);
```

Comment:
What happens if we are in the process of establishing a CASE session with another `OperationalDeviceProxy` when  we get here? Wouldn't that result in two separate, but similar CASE sessions (similar = to the same target) being present?